pfam_experiment.py

```diff
@@ -87,142 +127,171 @@ def main(_):
+		train_batches, train_indexes = create_pfam_batches(family_accessions=lens_knn_train_family_accessions,
+														   batch_size=FLAGS.batch_size,
+														   samples=FLAGS.lens_train_samples,
+														   epochs=measurement_epochs, 
+														   drop_remainder=True)
+
+		model = create_model(use_transformer=FLAGS.use_transformer, use_bert=FLAGS.use_bert, restore_transformer_dir=FLAGS.restore_transformer_dir,
+							 encoder_fn=encoder_fn, encoder_fn_kwargs=encoder_fn_kwargs, reduce_fn=reduce_fn, reduce_fn_kwargs=reduce_fn_kwargs, layers=layers, 
+							 output='prediction', encoder_fn_params=encoder_fn_params, reduce_fn_params=reduce_fn_params, predict_fn_params=predict_fn_params)
+
+		optimizer = train(model=model,
+	                      train_data=train_batches,
+	                      loss_fn=cross_entropy_loss,
+	                      loss_fn_kwargs=loss_fn_kwargs,
+	                      learning_rate=[FLAGS.encoder_lr, FLAGS.lens_lr, FLAGS.predictor_lr],
+	                      weight_decay=[FLAGS.encoder_wd, FLAGS.lens_wd, FLAGS.predictor_wd],
+	                      layers=layers)
+
+		trained_params = copy.deepcopy(optimizer.target.params)
+
+		results, preds = pfam_evaluate(predict_fn=optimizer.target,
+	                                   test_family_accessions=lens_knn_train_family_accessions,
+	                                   title=None,
+	                                   loss_fn_kwargs=loss_fn_kwargs,
+	                                   batch_size=FLAGS.batch_size)
+
+		lens_accuracy = results['accuracy']
+		datum['lens_accuracy' + '_measurement_' + str(i)] = lens_accuracy
+
+		lens_cross_entropy = float(results['cross_entropy'])
+		datum['lens_cross_entropy' + '_measurement_' + str(i)] = lens_cross_entropy
+
+		assert(embedding_optimizer.target.params.keys()==trained_params.keys()), 'Optimizer parameters do not match!'
+		for layer in embedding_optimizer.target.params.keys():
+			embedding_optimizer.target.params[layer] = trained_params[layer]
+
+		train_knn_results_trained_lens = pfam_nearest_neighbors_classification(encoder=embedding_optimizer.target, 
+					                                                               train_family_accessions=lens_knn_train_family_accessions, 
+					                                                               test_family_accessions=lens_knn_train_family_accessions,
+					                                                               batch_size=FLAGS.batch_size,
+					                                                               train_samples=1,
+					                                                               seed=1,
+					                                                               random_state=1)[0]
+		train_knn_accuracy_trained_lens = train_knn_results_trained_lens['1-nn accuracy']
+		datum['train_knn_accuracy_trained_lens_1_knn_train_samples' + '_measurement_' + str(i)] = train_knn_accuracy_trained_lens
+
+		for knn_train_samples in knn_train_samples_:
```

Comment:
can you add a helper function for knn evaluation and call it above and here. The risk of such code duplication is that they will be slightly different and not comparable. It also will improve the readability.